topologies/binding/cli.go

```diff
@@ -32,21 +32,38 @@ func newCLI(sc *ssh.Client) (*cli, error) {
 	return &cli{ssh: sc}, nil
 }
 
-func (c *cli) RunCommand(_ context.Context, cmd string) (binding.CommandResult, error) {
+func (c *cli) RunCommand(ctx context.Context, cmd string) (binding.CommandResult, error) {
 	sess, err := c.ssh.NewSession()
 	if err != nil {
 		return nil, fmt.Errorf("could not create session: %w", err)
 	}
 	defer sess.Close()
 
-	out, err := sess.CombinedOutput(cmd)
-	switch err.(type) {
-	case nil:
-		return &cmdResult{output: string(out)}, nil
-	case *ssh.ExitError, *ssh.ExitMissingError:
-		return &cmdResult{output: string(out), error: err.Error()}, nil
-	default:
-		return nil, err
+	outChan := make(chan *struct {
+		out string
+		err error
+	})
+
+	go func() {
```

Comment:
why introduce a goroutine?  could this introduce flakiness?  Wouldn't it be more simple to just keep this in the main process/thread as it was before?

Reply:
Sorry, I should clarify that the proposed change in DialCLI is unrelated to the main issue but seemed close enough in spirit to tack onto this PR. DialCLI as-is ignores the timeout value from the binding file (which is handled in the [crypto package](https://github.com/golang/crypto/blob/bb80217080b0e04c6e73e5dcd9f3a9bb11fe23f6/ssh/client.go#L177)).

I can remove it from this PR but it seemed like a relevant enough custodial change that I ended up including it. 

---

The issue seems to be the edge case where the ssh session connected to the DUT doesn't die, but hangs indefinitely. The `killall -STOP sshd` is just pausing the DUT's sshd in order to repro the client's behaviour.

> A test that triggers a reboot should expect this to occur and handle the timeout.

I agree that a test expecting an unresponsive sshd would need to handle the timeout. To me this means providing a context with a deadline to whatever request is being made, and deferring a cancellation in the thread making the call. Ideally this would mean passing down the ctx all the way to whatever dependency actually makes the request, however it is not forwarded further down the call stack beyond RunCommand. This is the reason I would like to handle it here. (i.e. sess.CombinedOutput doesn't accept context, and this is the function that hangs indefinitely when the DUT's sshd is unresponsive). The main thread will wait for the first of either: context timeout or CombinedOutput to return. 

As for handling it without a goroutine, I'm not too sure how it would be implemented. Maybe I am overlooking something?

Thanks for taking the time to review!